File: src/main/java/com/github/insanusmokrassar/iobject/realisations/ConcurrentSimpleIObject.java
package com.github.insanusmokrassar.iobject.realisations;

import com.github.insanusmokrassar.iobject.exceptions.ReadException;
import com.github.insanusmokrassar.iobject.exceptions.WriteException;
import com.github.insanusmokrassar.iobject.interfaces.IObject;

import java.util.Map;
import java.util.Set;

public class ConcurrentSimpleIObject<T> extends SimpleIObject<T>{

    public ConcurrentSimpleIObject(Map<String, T> from) {
        super(from);
    }

    public ConcurrentSimpleIObject(IObject<T> from) {
        super(from);
    }

    public ConcurrentSimpleIObject() {
        super();
    }

    @Override
    synchronized public void put(String key, T value) throws WriteException {
        super.put(key, value);
    }

    @Override
    synchronized public void putAll(Map<String, T> toPutMap) throws WriteException {
        super.putAll(toPutMap);
    }

    @Override
    synchronized public T get(String key) throws ReadException {
        return super.get(key);
    }

    @Override
    synchronized public void remove(String key) throws WriteException {
        super.remove(key);
    }

    @Override
    synchronized public Set<String> keys() {
        return super.keys();
    }

    @Override
    protected SimpleIObject<T> clone() throws CloneNotSupportedException {
        return new ConcurrentSimpleIObject<>(this);
    }
}

File: src/main/kotlin/com/github/insanusmokrassar/IObjectK/realisations/ConcurrentSimpleIObject.kt
package com.github.insanusmokrassar.IObjectK.realisations

import com.github.insanusmokrassar.IObjectK.exceptions.ReadException
import com.github.insanusmokrassar.IObjectK.exceptions.WriteException
import com.github.insanusmokrassar.IObjectK.interfaces.IObject

class ConcurrentSimpleIObject : SimpleIObject {

    constructor(from: Map<String, Any>) : super(from) {}

    constructor(from: IObject<Any>) : super(from) {}

    constructor() : super() {}

    @Synchronized @Throws(WriteException::class)
    override fun put(key: String, value: Any) {
        super.put(key, value)
    }

    @Synchronized @Throws(WriteException::class)
    override fun putAll(toPutMap: Map<String, Any>) {
        super.putAll(toPutMap)
    }

    @Synchronized @Throws(ReadException::class)
    override operator fun <T : Any>get(key: String): T {
        return super.get<T>(key)
    }

    @Synchronized @Throws(WriteException::class)
    override fun remove(key: String) {
        super.remove(key)
    }

    @Synchronized override fun keys(): Set<String> {
        return super.keys()
    }
}